matchzoo/models/anmm.py

```diff
@@ -0,0 +1,161 @@
+"""An implementation of aNMM Model."""
+import math
+import typing
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from matchzoo.dataloader import callbacks
+from matchzoo.engine import hyper_spaces
+from matchzoo.engine.base_model import BaseModel
+from matchzoo.engine.param import Param
+from matchzoo.engine.param_table import ParamTable
+from matchzoo.modules import Matching
+from matchzoo.utils import parse_activation
+
+
+class aNMM(BaseModel):
+    """
+    aNMM: Ranking Short Answer Texts with Attention-Based Neural Matching Model.
+
+    Examples:
+        >>> model = aNMM()
+        >>> model.params['embedding_output_dim'] = 300
+        >>> model.guess_and_fill_missing_params(verbose=0)
+        >>> model.build()
+
+    """
+
+    @classmethod
+    def get_default_params(cls) -> ParamTable:
+        """:return: model default parameters."""
+        params = super().get_default_params(with_embedding=True)
+        params.add(Param(name='mask_value', value=0,
+                         desc="The value to be masked from inputs."))
+        params.add(Param(name='num_bins', value=200,
+                         desc="Integer, number of bins."))
+        params.add(Param(name='hidden_sizes', value=[100],
+                         desc="Number of hidden size for each hidden layer"))
+        params.add(Param(name='activation', value='relu',
+                         desc="The activation function."))
+
+        params.add(Param(
+            'dropout_rate', 0.0,
+            hyper_space=hyper_spaces.quniform(
+                low=0.0, high=0.8, q=0.01),
+            desc="The dropout rate."
+        ))
+        return params
+
+    def build(self):
+        """
+        Build model structure.
+
+        aNMM: Ranking Short Answer Texts with Attention-Based Neural Matching Model.
+        """
+        self.embedding = self._make_default_embedding_layer()
+
+        # QA Matching
+        self.matching = Matching(matching_type='dot', normalize=True)
+
+        # Value-shared Weighting
+        activation = parse_activation(self._params['activation'])
+        in_hidden_size = [
+            self._params['num_bins'],
+            *self._params['hidden_sizes']
+        ]
+        out_hidden_size = [
+            *self._params['hidden_sizes'],
+            1
+        ]
+
+        hidden_layers = [
+            nn.Sequential(
+                nn.Linear(in_size, out_size),
+                activation
+            )
+            for in_size, out_size, in zip(
+                in_hidden_size,
+                out_hidden_size
+            )
+        ]
+        self.hidden_layers = nn.Sequential(*hidden_layers)
+
+        # Query Attention
+        self.q_attention = nn.Linear(self._params['embedding_output_dim'], 1, bias=False)
```

Comment:
Use `Attention` module to rewrite. And remove `math` package.
https://github.com/NTMC-Community/MatchZoo-py/blob/f4c28583b7cd076dd128ea0916ed49a372b98f14/matchzoo/modules/attention.py#L9-L37